<?xml version="1.0"?>
<launch>
    <!-- Different arguments for each vehicle -->

    <arg name="namespace" default="ugv1"/>

    <arg name="ugv_type" default="0"/> <!-- 0 mecanum, 1 unicycle, 2 ackermann -->
    <arg name="init_x_East_metre" default="0.0"/>
    <arg name="init_y_North_metre" default="0.0"/>
    <arg name="init_z_Up_metre" default="0.0"/>
    <arg name="init_roll_deg" default="0.0"/>
    <arg name="init_pitch_deg" default="0.0"/>
    <arg name="init_yaw_deg" default="0.0"/>

    <arg name="v_front_max_m_s" default="1.5"/> <!-- m/s -->
    <arg name="v_left_max_m_s" default="1.5"/> <!-- m/s -->
    <arg name="omega_max_deg_s" default="90.0"/> <!-- deg/s -->
    <arg name="min_turning_radius_m" default="0.2"/> <!-- only for ackermann -->

    <arg name="linear_vel_scale" default="1.0"/> <!-- cmd_vel.linear.xy to linear velocity(m/s)  -->
    <arg name="angular_vel_scale" default="1.0"/> <!-- cmd_vel.angular.z to angular velocity(rad/s)  -->
    <arg name="use_imu_orientation" default="false"/> <!-- if true, the pose orientation will be imu(mag) orientation, which may not be aligned with mocap/uwb world frame -->

    <arg name="use_sim_time" default="true"/> <!-- Note that 'use_sim_time' param is global. All nodes in all launch files will use sim time if use_sim_time is set true in any launch file -->
    <arg name="open_rviz" default="true"/>
    
    <!-- if not use sim time, no need to use nodelet as there is no accelerating -->
    <arg name="use_nodelet_manager" value="$(arg use_sim_time)"/>
    <!-- <arg name="use_nodelet_manager" value="false"/> -->
    <param name="/use_sim_time" value="$(arg use_sim_time)"/>

<group ns="$(arg namespace)">
    <group if="$(arg use_nodelet_manager)">  <!-- /nodelet_manager should be lauched by sim_clock.launch -->
        <node pkg="nodelet" type="nodelet" name="wheeltec_ugv_sim_nodelet_$(arg namespace)" args="load wheeltec_ugv_sim_nodelet /nodelet_manager" output="screen">
            <param name="ugv_type" value="$(arg ugv_type)"/>
            <param name="init_x_East_metre" value="$(arg init_x_East_metre)"/>
            <param name="init_y_North_metre" value="$(arg init_y_North_metre)"/>
            <param name="init_z_Up_metre" value="$(arg init_z_Up_metre)"/>
            <param name="init_roll_deg" value="$(arg init_roll_deg)"/>
            <param name="init_pitch_deg" value="$(arg init_pitch_deg)"/>
            <param name="init_yaw_deg" value="$(arg init_yaw_deg)"/>
            <param name="v_front_max_m_s" value="$(arg v_front_max_m_s)"/>
            <param name="v_left_max_m_s" value="$(arg v_left_max_m_s)"/>
            <param name="omega_max_deg_s" value="$(arg omega_max_deg_s)"/>
            <param name="min_turning_radius_m" value="$(arg min_turning_radius_m)"/>
            <param name="linear_vel_scale" value="$(arg linear_vel_scale)"/>
            <param name="angular_vel_scale" value="$(arg angular_vel_scale)"/>
            <param name="use_imu_orientation" value="$(arg use_imu_orientation)"/>
        </node>
    </group>

    <group unless="$(arg use_nodelet_manager)">  <!--launched as a normal ROS node -->
        <node pkg="nodelet" type="nodelet" name="wheeltec_ugv_sim_nodelet_$(arg namespace)" args="standalone wheeltec_ugv_sim_nodelet" output="screen">
            <param name="ugv_type" value="$(arg ugv_type)"/>
            <param name="init_x_East_metre" value="$(arg init_x_East_metre)"/>
            <param name="init_y_North_metre" value="$(arg init_y_North_metre)"/>
            <param name="init_z_Up_metre" value="$(arg init_z_Up_metre)"/>
            <param name="init_roll_deg" value="$(arg init_roll_deg)"/>
            <param name="init_pitch_deg" value="$(arg init_pitch_deg)"/>
            <param name="init_yaw_deg" value="$(arg init_yaw_deg)"/>
            <param name="v_front_max_m_s" value="$(arg v_front_max_m_s)"/>
            <param name="v_left_max_m_s" value="$(arg v_left_max_m_s)"/>
            <param name="omega_max_deg_s" value="$(arg omega_max_deg_s)"/>
            <param name="min_turning_radius_m" value="$(arg min_turning_radius_m)"/>
            <param name="linear_vel_scale" value="$(arg linear_vel_scale)"/>
            <param name="angular_vel_scale" value="$(arg angular_vel_scale)"/>
            <param name="use_imu_orientation" value="$(arg use_imu_orientation)"/>
        </node>
    </group>
</group>

<!-- Visualization -->
<include file="$(find ugv_sim)/launch/ugv_visualizer_single.launch">
    <arg name="namespace" value="$(arg namespace)"/>
    <arg name="open_rviz" value="$(arg open_rviz)"/>
</include>


</launch>
